src/lib/gradientParser.ts

```diff
@@ -0,0 +1,361 @@
+  type: "color-hint"
+  hint: Length
+}
+
+export type LinearGradient = {
+  type: "linear-gradient" | "repeating-linear-gradient"
+  gradientLine: SideOrCornerGradientLine | AngleGradientLine
+  colorStops: (ColorStop | ColorHint)[]
+}
+
+export type ConicGradient = {
+  type: "conic-gradient" | "repeating-conic-gradient"
+  // Normalized to degrees
+  angle?: number
+  position: string
+  colorStops: (AngularColorStop | ColorHint)[]
+}
+
+export type RadialGradient = {
+  type: "radial-gradient" | "repeating-radial-gradient"
+  endingShape: "circle" | "ellipse"
+  size: "closest-corner" | "closest-side" | "farthest-corner" | "farthest-side" | Length[]
+  position: string
+  colorStops: (ColorStop | ColorHint)[]
+}
+
+const ANGLE_UNITS = ["deg", "turn", "rad", "grad"]
+const ANGLE_OR_PERCENTAGE_UNITS = [...ANGLE_UNITS, "%"]
+
+export function parseGradient(css: string): GradientNode[] {
+  const parsed = parse(css)
+  const gradients = splitCommaArgs(parsed.nodes)
+  return gradients
+    .filter(
+      (nodes) =>
+        // background-image can include non-gradient values, we can ignore them
+        nodes.length === 1 && nodes[0].type === "function" && nodes[0].value.includes("-gradient")
+    )
+    .map(([node]) => {
+      const args = splitCommaArgs((node as FunctionNode).nodes)
+      // strip vendor prefixes
+      const type = node.value.replace(/^-webkit-/, "")
+
+      if (type.includes("linear-gradient")) {
+        return parseLinearGradient(type, args)
+      } else if (type.includes("radial-gradient")) {
+        return parseRadialGradient(type, args)
+      } else if (type.includes("conic-gradient")) {
+        return parseConicGradient(type, args)
+      }
+      throw new Error("Unsupported gradient: " + stringify(node))
+    })
+}
+
+function parseLinearGradient(type: string, args: parse.Node[][]) {
+  const ret: Partial<LinearGradient> = {
+    type: type as LinearGradient["type"]
+  }
+
+  if (args[0][0].value === "to") {
+    ret.gradientLine = {
+      type: "side-or-corner",
+      value: args
+        .shift()!
+        .slice(1)
+        .map((it) => it.value)
+        .join(" ") as SideOrCornerGradientLine["value"]
+    }
+  } else {
+    const first = toUnit(args[0][0], "deg")
+    if (first && ANGLE_UNITS.includes(first.unit)) {
+      ret.gradientLine = {
+        type: "angle",
+        value: toDegrees(first)
+      }
+      args.shift()
+    } else {
+      // default gradient line if none is specified
+      ret.gradientLine = {
+        type: "angle",
+        value: 180
+      }
+    }
+  }
+
+  ret.colorStops = args.map(toColorStopOrHint)
+
+  return ret as LinearGradient
+}
+
+function parseRadialGradient(type: string, args: parse.Node[][]) {
+  const ret: Partial<RadialGradient> = {
+    type: type as RadialGradient["type"],
+    endingShape: "ellipse",
+    size: "farthest-corner",
+    position: "center"
+  }
+
+  let hasOptionalArg = false
+  loop: for (let i = 0; i < args[0].length; i++) {
```

Comment:
@gluck this loop kinda bugs me. I feel like there most be an easier way.

I see why you did it trying to make sure the keywords are consumed because they can come in any order.